### PR TITLE
feat(payment): INT-4205 Add Moneris strategy

### PR DESCRIPTION
**Pull request by FranciscoOnt**
## What? [INT-4205](https://jira.bigcommerce.com/browse/INT-4205)
Create Moneris payment strategy which mounts an iframe and sends events to the iframe and listens to evens sent from the iframe

## Why?
This is the way its implementation works and its based on [their documentation](https://developer.moneris.com/Documentation/NA/E-Commerce%20Solutions/Hosted%20Solutions/Hosted%20Tokenization) for the Hosted tokenization 

## Sibling PRs
https://github.com/bigcommerce/checkout-js/pull/563

## Testing / Proof
image

@bigcommerce/checkout @bigcommerce/payments @bigcommerce/apex-integrations

### CODE REVIEW

**Review by JoseAlvarezHdz — COMMENTED** [review]
Its looking good just a small change!

File: src/payment/strategies/moneris/moneris.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+export default interface MonerisStylingProps {
+    cssBody?: string;
+    cssTextbox?: string;
+    cssTextboxPan?: string;
+    cssTextboxExpiry?: string;
+    csstexboxCvd?: string;
+}
+
+export interface MonerisResponseData {
+  responseCode: string[];
```
Comment (JoseAlvarezHdz) [inline src/payment/strategies/moneris/moneris.ts:10]: Please fix indentation

File: src/payment/strategies/moneris/moneris-payment-initialize-options.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import MonerisStylingProps from './moneris';
+
+export default interface MonerisaymentInitializeOptions {
+    /**
+     * The ID of a container which the Moneris iframe component should be mounted
+     */
+  containerId: string;
```
Comment (JoseAlvarezHdz) [inline src/payment/strategies/moneris/moneris-payment-initialize-options.ts:7]: Fix Indentation please

**Review by JoseAlvarezHdz — APPROVED** [review]
LGTM, just some indentation to fix

File: src/payment/strategies/moneris/moneris-payment-strategy.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        const { isStoreCreditApplied: useStoreCredit } = this._store.getState().checkout.getCheckoutOrThrow();
+
+        if (useStoreCredit !== undefined) {
+            await this._store.dispatch(this._storeCreditActionCreator.applyStoreCredit(useStoreCredit));
+        }
+
+        await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
+
+        const nonce = await new Promise<string | undefined>((resolve, reject) => {
+            const frameref = this._iframe?.contentWindow;
```
Comment (capsula4) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:63]: should you check that `this._iframe?` is defined? and if not, throw an error that the strategy hasn't been initialized

File: src/payment/strategies/moneris/moneris-payment-strategy.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+        return this._store.getState();
+    }
+
+    finalize(): Promise<InternalCheckoutSelectors> {
+        return Promise.reject(new OrderFinalizationNotRequiredError());
+    }
+
+    deinitialize(): Promise<InternalCheckoutSelectors> {
+        if (this._windowEventListener) {
```
Comment (capsula4) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:91]: should you remove the iframe as well?

File: src/payment/strategies/moneris/moneris-payment-strategy.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        const cssBody = style?.cssBody ?? 'background:white;';
+        const cssTextbox = style?.cssTextbox ?? 'border-width:2px;';
+        const csstextboxPan = style?.cssTextboxPan ?? 'width:140px;';
+        const cssTextboxExpiry = style?.cssTextboxExpiry ?? 'width:40px;';
+        const csstexboxCvd = style?.csstexboxCvd ?? 'width:40px';
+
+        iframe.name = IFRAME_NAME;
+        iframe.id = IFRAME_NAME;
+
+        this._monerisURL = `https://${ testMode ? 'esqa' : 'www3' }.moneris.com/HPPtoken/index.php`;
```
Comment (capsula4) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:117]: does this need to be a class property?

File: src/payment/strategies/moneris/moneris-payment-strategy.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        if (useStoreCredit !== undefined) {
+            await this._store.dispatch(this._storeCreditActionCreator.applyStoreCredit(useStoreCredit));
+        }
+
+        await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
+
+        const nonce = await new Promise<string | undefined>((resolve, reject) => {
+            const frameref = this._iframe?.contentWindow;
+
+            if (!this._monerisURL) {
```
Comment (capsula4) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:65]: maybe this should be replaced by the iframe property instead?

File: src/payment/strategies/moneris/moneris-payment-strategy.ts (hunk 4 above)
Comment (capsula4) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:91]: should you also remove the iframe and the monerisUrl?

File: src/payment/strategies/moneris/moneris-payment-strategy.ts (hunk 5 above)
Comment (capsula4) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:117]: should this be initialized in the constructor?

File: src/payment/strategies/moneris/moneris-payment-strategy.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+        if (!monerisOptions) {
+            throw new InvalidArgumentError('Unable to initialize payment because "options.moneris" argument is not provided.');
+        }
+
+        if (!initializationData?.profileId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        this._iframe = this._createIframe(monerisOptions.containerId, initializationData.profileId, !!config.testMode);
```
Comment (capsula4) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:42]: maybe you can check that the iframe hasn't been created?


if (this._iframe) {
 throw new Error('Already initalized');
}

?

File: src/payment/strategies/moneris/moneris-payment-initialize-options.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import MonerisStylingProps from './moneris';
+
+export default interface MonerisaymentInitializeOptions {
+    /**
+     * The ID of a container which the Moneris iframe component should be mounted
```
Comment (icatalina) [inline src/payment/strategies/moneris/moneris-payment-initialize-options.ts:5]:
```suggestion
     * The ID of a container where the Moneris iframe component should be mounted
```

File: src/payment/strategies/moneris/moneris-payment-strategy.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        iframe.id = IFRAME_NAME;
+
+        iframe.src = `${ this._monerisURL(testMode) }?id=${profileId}&pmmsg=true&css_body=${cssBody}&css_textbox=${cssTextbox}&css_textbox_pan=${csstextboxPan}&enable_exp=1&css_textbox_exp=${cssTextboxExpiry}&enable_cvd=1&css_textbox_cvd=${csstexboxCvd}&display_labels=1`;
+
+        container.appendChild(iframe);
+
+        return iframe;
+    }
+
+    private _handleMonerisResponse(response: MessageEvent, resolve: (nonce: string) => void, reject: (reason: Error) => void): void {
```
Comment (icatalina) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:134]: It is a bit awkward to pass the responsibility of resolving/rejecting the promise into this function. Might be better to keep this function just about handling the response and move the `resolve`/`reject` part to the caller.

```ts
private _handleMonerisResponse(response: MessageEvent): string {
```

```ts
this._windowEventListener = (response: MessageEvent) => {
  try {
    resolve(this._handleMonerisResponse(response));
  } catch (error) {
    reject(error);
  }
};
```

File: src/payment/strategies/moneris/moneris-payment-strategy.spec.ts [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+            eventEmitter.emit('message', { data: '{\"responseCode\":[\"400\"],\"errorMessage\":\"expected error message\",\"dataKey\":\"ABC123\",\"bin\":\"1234\"}'});
+
+            await expect(promise).rejects.toThrow(new Error('expected error message'));
+
+            expect(await strategy.deinitialize()).toEqual(store.getState());
+            expect(window.removeEventListener).toHaveBeenCalledWith('message', expect.any(Function));
+        });
+
+        it('deinitializes strategy and removes the iframe is it exists', async () => {
```
Comment (icatalina) [inline src/payment/strategies/moneris/moneris-payment-strategy.spec.ts:253]:
```suggestion
        it('deinitializes strategy and removes the iframe if it exists', async () => {
```

File: src/payment/strategies/moneris/moneris-payment-strategy.spec.ts [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+            const expectedPayment = {
+                methodId: 'moneris',
+                paymentData: {
+                    nonce: 'ABC123',
+                },
+            };
+            checkoutMock.isStoreCreditApplied = true;
+
+            await strategy.initialize(initializeOptions);
+            const promise =  strategy.execute(payload, options);
```
Comment (icatalina) [inline src/payment/strategies/moneris/moneris-payment-strategy.spec.ts:199]:
```suggestion
            const promise = strategy.execute(payload, options);
```

**Review by icatalina — APPROVED** [review]
Hey @FranciscoOnt!

Looking good, I just added a few 🍹 comments.

**Pull request merged**